crates/driver/src/domain/competition/order/mod.rs

```diff
@@ -133,53 +161,39 @@ impl Order {
         matches!(self.kind, Kind::Liquidity)
     }
 
-    /// The available sell amount for the order.
+    /// Returns the order's available amounts to be passed to a solver engine.
     ///
-    /// This returns the maximum partial sell amount that can be used given past
-    /// partial executions and available balance.
-    pub fn available_sell(&self) -> eth::Asset {
-        match self.partial {
-            Partial::Yes { available } => eth::Asset {
-                token: self.sell.token,
-                amount: util::math::mul_ratio(self.sell.amount.0, available.0, self.target().0)
-                    .unwrap_or_default()
-                    .into(),
+    /// See [`Available`] for more details.
+    pub fn available(&self, weth: eth::WethAddress) -> Available {
+        let mut amounts = Available {
+            sell: self.sell,
+            buy: eth::Asset {
+                token: self.buy.token.wrap(weth),
+                amount: self.buy.amount,
             },
-            Partial::No => self.sell,
-        }
-    }
+            fee: self.fee,
+        };
 
-    /// The available buy amount for the order.
-    ///
-    /// This returns the minimum partial buy amount that can be used given past
-    /// partial executions and available balance.
-    pub fn available_buy(&self) -> eth::Asset {
-        match self.partial {
-            Partial::Yes { available } => eth::Asset {
-                token: self.buy.token,
-                amount: util::math::mul_ratio_ceil(self.buy.amount.0, available.0, self.target().0)
-                    .unwrap_or_default()
-                    .into(),
-            },
-            Partial::No => self.buy,
+        let available = match self.partial {
+            Partial::Yes { available } => available,
+            Partial::No => return amounts,
+        };
+        let target = self.target();
+
+        for amount in [
+            &mut amounts.sell.amount.0,
+            &mut amounts.fee.user.0,
+            &mut amounts.fee.solver.0,
+        ] {
+            *amount = util::math::mul_ratio(*amount, available.0, target.0).unwrap_or_default();
```

Comment:
Given that target cannot be null I wonder if just unwrapping is fine here.

Reply:
I kept it as `unwrap_or_default()` since we have no type guarantees that it is non-zero (maybe we should use non-zero U256 types in this code as well!).

Also, conceptually, an order selling 0 would "scale to 0", so I think the behaviour makes sense mathematically as well.